File: main/4sum/4sum.java
class Solution {
    public List<List<Integer>> fourSum(int[] nums, int target) {
        final List<List<Integer>> ret = new ArrayList<>();
        if (nums.length < 4) return ret;

        final int gstop = nums.length - 3, hstop = nums.length - 2;
        Arrays.sort(nums);

        for (int g = 0; g != gstop; ) {
            final int gval = nums[g], hij_target = target - gval;

            for (int h = g + 1; h != hstop; ) {
                final int hval = nums[h], ij_target = hij_target - hval;

                for (int i = h + 1, j = nums.length - 1; i < j; ) {
                    final int ival = nums[i], jval = nums[j];
                    final int ij_sum = ival + jval;

                    if (ij_sum == ij_target)
                        ret.add(Arrays.asList(gval, hval, ival, jval));

                    if (ij_sum <= ij_target)
                        while (nums[i] == ival && i < j) ++i;

                    if (ij_sum >= ij_target)
                        while (nums[j] == jval && i < j) --j;
                }

                while (nums[h] == hval && h != hstop) ++h;
            }

            while (nums[g] == gval && g != gstop) ++g;
        }

        return ret;
    }
}

File: main/4sum/4sum.py
class Solution:
    def fourSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        ret = []

        n = len(nums)
        if n < 4:
            return ret

        nums.sort()
        gval = None

        for g in range(n - 3):
            if nums[g] == gval:
                continue

            gval = nums[g]
            hij_target = target - gval
            hval = None

            for h in range(g + 1, n - 2):
                if nums[h] == hval:
                    continue

                hval = nums[h]
                ij_target = hij_target - hval
                i = h + 1
                j = n - 1

                while i < j:
                    ival = nums[i]
                    jval = nums[j]
                    ij_sum = ival + jval

                    if ij_sum == ij_target:
                        ret.append((gval, hval, ival, jval))

                    if ij_sum <= ij_target:
                        while nums[i] == ival and i < j:
                            i += 1

                    if ij_sum >= ij_target:
                        while nums[j] == jval and i < j:
                            j -= 1

        return ret
